rig-core/src/providers/gemini/completion.rs

```diff
@@ -0,0 +1,607 @@
+                            })
+                            .collect()
+                    }),
+                required: obj.get("required").and_then(|v| v.as_array()).map(|arr| {
+                    arr.iter()
+                        .filter_map(|v| v.as_str().map(String::from))
+                        .collect()
+                }),
+                items: obj
+                    .get("items")
+                    .map(|v| Box::new(v.clone().try_into().unwrap())),
+            })
+        } else {
+            Err(CompletionError::ResponseError(
+                "Expected a JSON object for Schema".into(),
+            ))
+        }
+    }
+}
+
+#[derive(Debug, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct GenerateContentRequest {
+    pub contents: Vec<Content>,
+    pub tools: Option<Vec<Tool>>,
+    pub tool_config: Option<ToolConfig>,
+    pub generation_config: Option<GenerationConfig>,
+    pub safety_settings: Option<Vec<SafetySetting>>,
+    pub system_instruction: Option<String>,
+    // cachedContent: Optional<String>
+}
+
+#[derive(Debug, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct Tool {
+    pub function_declaration: FunctionDeclaration,
+    pub code_execution: Option<CodeExecution>,
+}
+
+#[derive(Debug, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct FunctionDeclaration {
+    pub name: String,
+    pub description: String,
+    pub parameters: Option<Vec<Schema>>,
+}
+
+#[derive(Debug, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct ToolConfig {
+    pub schema: Option<Schema>,
+}
+
+#[derive(Debug, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct CodeExecution {}
+
+#[derive(Debug, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct SafetySetting {
+    pub category: HarmCategory,
+    pub threshold: HarmBlockThreshold,
+}
+
+#[derive(Debug, Serialize)]
+#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
+pub enum HarmBlockThreshold {
+    HarmBlockThresholdUnspecified,
+    BlockLowAndAbove,
+    BlockMediumAndAbove,
+    BlockOnlyHigh,
+    BlockNone,
+    Off,
+}
+
+// =================================================================
+// Rig Implementation Types
+// =================================================================
+
+#[derive(Clone)]
+pub struct CompletionModel {
+    client: Client,
+    pub model: String,
+}
+
+impl CompletionModel {
+    pub fn new(client: Client, model: &str) -> Self {
+        Self {
+            client,
+            model: model.to_string(),
+        }
+    }
+}
+
+impl completion::CompletionModel for CompletionModel {
+    type Response = GenerateContentResponse;
+
+    async fn completion(
+        &self,
+        mut completion_request: CompletionRequest,
+    ) -> Result<completion::CompletionResponse<GenerateContentResponse>, CompletionError> {
+        // QUESTION: Why do Anthropic/openAi implementation differ here? OpenAI adds the preamble but Anthropic does not.
```

Comment:
For Anthropic, the Preamble is set as the `system` API field. The other ones differ add it as a `system` message in the chat history. Differences in API design.

Reply:
Made the change! This is `system_instruction` in the request object.